HTTP/2: fixed flushing early hints.

arut opened the pull request:
Previously, when using HTTP/2 over SSL, an early hints HEADERS frame was queued in SSL buffer, but might not be immediately flushed.  This resulted in a delay of early hints delivery until the main response header was sent.

The fix is to set the flush flag for the early hints HEADERS frame buffer.

Reported in https://github.com/nginx/nginx/pull/326#issuecomment-3111364338 and #807.

arut committed:
HTTP/2: fixed flushing early hints over SSL.

Previously, when using HTTP/2 over SSL, an early hints HEADERS frame was
queued in SSL buffer, and might not be immediately flushed.  This resulted
in a delay of early hints delivery until the main response was sent.

The fix is to set the flush flag for the early hints HEADERS frame buffer.

pluknet commented:
> HTTP/2: fixed flushing early hints.

HTTP/2: flushing of SSL buffered early hints ?

> Previously, when using HTTP/2 over SSL, an early hints HEADERS frame was
> queued in SSL buffer, but might not be immediately flushed.  This resulted

s/, but/ and/, imho

> in a delay of early hints delivery until the main response header was sent.

Only DATA frames are actually (always) flushed.

In this regard, looking at the code, I find it quite logical to consistently set
this flag in the same function that is used to prepare the frame buffer, that is,
ngx_http_v2_filter_get_data_frame() and ngx_http_v2_create_headers_frame().

The latter one is reused to construct main response headers and trailers, so
an additional flag could be introduced to signal that flush is needed.
It might look as follows:

```diff
@@ -916,6 +917,7 @@ ngx_http_v2_create_headers_frame(ngx_http_request_t *r, u_char *pos,
         }
 
         b->last_buf = fin;
+        b->flush = flush;
         cl->next = NULL;
         frame->last = cl;
 
```

arut commented:
> ```diff
> ngx_http_v2_create_headers_frame
> ```

This was my first version. I added a `flush` argument to `ngx_http_v2_create_headers_frame()` and passed different values there. The patch is much bigger, but it's not obvious to me that it all makes sense for such a small feature. I will post this version again if that's your concern.

pluknet commented:
> > ```diff
> > ngx_http_v2_create_headers_frame
> > ```
> 
> This was my first version. I added a `flush` argument to `ngx_http_v2_create_headers_frame()` and passed different values there. The patch is much bigger, but it's not obvious to me that it all makes sense for such a small feature. I will post this version again if that's your concern.

Yes, please.
IMHO, this would make code more clean for the above reasons.

arut commented:
> HTTP/2: flushing of SSL buffered early hints ?

Too many details IMHO. I suggest this version

> HTTP/2: fixed flushing early hints over SSL.

Updated the commit log. Updated the patch.
